### PR TITLE
Feature: enables printing out the bandgap at each SCF step

### LINKED ISSUE #1667: Band gap should be printed in SCF loops
### Background

Band gap is related to the choosing of parameters "smearing_type" and "smearing_sigma".
It should be output in SCF loops for user judging if the input parameters should be modified.

### Describe the solution you'd like

In "OUT.ABACUS/running_scf.log", band gap value can be printed after calculating occupied weights.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by ouqi0711**

### CODE REVIEW

**Review by dyzheng — CHANGES_REQUESTED** [review]
In input.cpp:Bcast() function, every parameter should be casted to other processes.

**Review by dyzheng — COMMENTED** [review]
In src_io/write_input.cpp:Print() function, every parameter should be printed to OUT.ABACUS/INPUT file.

**Review by dyzheng — COMMENTED** [review]
Please modify at least one test case in test/integral/, just set "out_bandgap 1" to coverage these code, I guess correctness will not change with this modification.

**Comment by dyzheng** [comment]
The CI error will be fixed in #1698 , it is not caused by this PR.

**Comment by ouqi0711** [comment]
> Please modify at least one test case in test/integral/, just set "out_bandgap 1" to coverage these code, I guess correctness will not change with this modification.

Is that necessary to modify the result.ref file so as to capture the bandgap result? Or it's fair enough just setting out_bandgap =1 in an INPUT?

**Comment by dyzheng** [comment]
> > Please modify at least one test case in test/integral/, just set "out_bandgap 1" to coverage these code, I guess correctness will not change with this modification.
> 
> Is that necessary to modify the result.ref file so as to capture the bandgap result? Or it's fair enough just setting out_bandgap =1 in an INPUT?

I think there is no need to modify result.ref now.

File: source/input.cpp [hunk 1]
```diff
@@ -1076,6 +1078,10 @@ bool Input::Read(const std::string &fn)
         {
             read_bool(ifs, out_dm1);
         }
+        else if (strcmp("out_bandgap", word) == 0) // for bandgap printing
+        {
+            read_value(ifs, out_bandgap);
```
Comment (hongriTianqi) [inline source/input.cpp]:
```suggestion
            read_bool(ifs, out_bandgap);
```

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/src_io/write_input.cpp [hunk 2]
```diff
@@ -177,6 +177,7 @@ void Input::Print(const std::string &fn) const
     ModuleBase::GlobalFunc::OUTP(ofs, "relax_scale_force", relax_scale_force, "controls the size of the first CG step if relax_new is true");
     ModuleBase::GlobalFunc::OUTP(ofs, "out_level", out_level, "ie(for electrons); i(for ions);");
     ModuleBase::GlobalFunc::OUTP(ofs, "out_dm", out_dm, ">0 output density matrix");
+    ModuleBase::GlobalFunc::OUTP(ofs, "out_bandgap", out_bandgap, ">0 print out bandgap");
```
Comment (hongriTianqi) [inline source/src_io/write_input.cpp]:
```suggestion
    ModuleBase::GlobalFunc::OUTP(ofs, "out_bandgap", out_bandgap, "if true, print out bandgap");
```

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Review by dyzheng — APPROVED** [review]

**Review by hongriTianqi — APPROVED** [review]

**Pull request merged**